core/primitives/src/types.rs

```diff
@@ -136,6 +136,50 @@ pub type RawStateChangesList = Vec<(StateChangeCause, Option<Vec<u8>>)>;
 /// key that was updated -> list of updates with the corresponding indexing event.
 pub type RawStateChanges = std::collections::BTreeMap<Vec<u8>, RawStateChangesList>;
 
+#[derive(Debug)]
+pub enum StateChangeKind {
+    AccountTouched { account_id: AccountId },
+    AccessKeyTouched { account_id: AccountId },
+    DataTouched { account_id: AccountId },
+    CodeTouched { account_id: AccountId },
+}
+
+pub type StateChangesKinds = Vec<StateChangeKind>;
+
+#[easy_ext::ext(StateChangesKindsExt)]
+impl StateChangesKinds {
+    pub fn from_changes<K: AsRef<[u8]>>(
+        raw_changes: &mut dyn Iterator<Item = Result<(K, RawStateChangesList), std::io::Error>>,
+    ) -> Result<StateChangesKinds, std::io::Error> {
+        let mut changes = Self::new();
+
+        for raw_change in raw_changes {
+            let (key, _) = raw_change?;
+
+            // WARNING: The order of checks is important as some keys are an extension of the others
+            // (e.g. KeyForData = KeyForAccount + separator + data-key).
+            let state_change_kind = if let Ok(account_id) = KeyForData::parse_account_id(&key) {
```

Comment:
This logic will break my changes to hash account_id in the trie: https://github.com/nearprotocol/nearcore/pull/2244

Because the key will only contain the hash of the account_id instead of the account_id itself. As part of my change I'm adding a new column and a record that maps back the `hash(account_id)` to `account_id`. But it's unclear how to query it from the current function. Basically, we'll have to do a state_read read for every unique hash that you get here

Reply:
I guess, I should change the stored change value to include StoreChangeKind, and avoid parsing the key altogether.

Reply:
Well, we still need to derive StoreChangeKind from the key, but during the saving stage, instead of the reading stage.